services/provider/server/server.go

```diff
@@ -1668,36 +1679,47 @@ func (s *OCSProviderServer) appendVolumeSnapshotClassKubeResources(
 			)
 		}
 	}
-	for i := range consumer.Spec.VolumeSnapshotClasses {
-		var snapshotClass *snapapi.VolumeSnapshotClass
-		var err error
-		snapshotClassName := consumer.Spec.VolumeSnapshotClasses[i].Name
-		vscGen := vscMap[snapshotClassName]
-		if vscGen != nil {
-			snapshotClass = vscGen()
-			snapshotClass.Name = snapshotClassName
-		} else {
-			snapshotClass, err = util.VolumeSnapshotClassFromExisting(
-				ctx,
-				s.client,
-				snapshotClassName,
-				consumer,
-				consumerConfig,
-				rbdStorageId,
-				cephFsStorageId,
-				cephFsStorageId,
-			)
-		}
-		if kerrors.IsNotFound(err) {
-			klog.Warningf("VolumeSnapshotClass with name %s doesn't exist in the cluster", snapshotClassName)
-		} else if errors.Is(err, util.UnsupportedDriver) {
-			klog.Warningf("Encountered unsupported driver in volume snapshot class %s", snapshotClassName)
-		} else if snapshotClass == nil {
-			klog.Warningf("The name %s does not points to a builtin or an existing volumesnapshot class, skipping", snapshotClassName)
-		} else if snapshotClass.Labels[util.ExternalClassLabelKey] == "true" {
-			klog.Warningf("The snapshot class %s is an external storage class, skipping", snapshotClassName)
-		} else {
-			kubeResources = append(kubeResources, snapshotClass)
+
+	kubeObjCache := map[string]*snapapi.VolumeSnapshotClass{}
+	for destName, srcName := range util.GetClassNamesMapping(consumer.Spec.VolumeSnapshotClasses) {
+		var kubeObj *snapapi.VolumeSnapshotClass
+		if kubeObj = kubeObjCache[srcName]; kubeObj == nil {
+			var snapshotClass *snapapi.VolumeSnapshotClass
+			var err error
+			vscGen := vscMap[srcName]
+			if vscGen != nil {
+				snapshotClass = vscGen()
+				snapshotClass.Name = srcName
+			} else {
+				snapshotClass, err = util.VolumeSnapshotClassFromExisting(
+					ctx,
+					s.client,
+					srcName,
+					consumer,
+					consumerConfig,
+					rbdStorageId,
+					cephFsStorageId,
+					cephFsStorageId,
+				)
+			}
+			if kerrors.IsNotFound(err) {
+				klog.Warningf("VolumeSnapshotClass with name %s doesn't exist in the cluster", srcName)
+			} else if errors.Is(err, util.UnsupportedDriver) {
+				klog.Warningf("Encountered unsupported driver in volume snapshot class %s", srcName)
+			} else if snapshotClass == nil {
+				klog.Warningf("The name %s does not points to a builtin or an existing volumesnapshot class, skipping", srcName)
+			} else if snapshotClass.Labels[util.ExternalClassLabelKey] == "true" {
+				klog.Warningf("The snapshot class %s is an external storage class, skipping", srcName)
+			} else {
+				kubeObj = snapshotClass
+				kubeObjCache[srcName] = snapshotClass
+			}
+			if kubeObj != nil {
+				distObj := &snapapi.VolumeSnapshotClass{}
+				kubeObj.DeepCopyInto(distObj)
+				distObj.Name = destName
+				kubeResources = append(kubeResources, distObj)
+			}
```

Comment:
This code is almost identical to the code for storageclasses, why isn't this logic part of the generic code?

Reply:
it uses a non-generic mapping `vscMap` and non-generic func `util.*ClassFromExisting`.

Reply:
But that is just how you decided to write the code. I don't see any reason why this cannot work in a generic way.
One where you get the a function producing the actual class as a `client.Object` as an argument. 
the mapping is always `map[string]string` and the result is always []client.Object`